### DISCUSSION TITLE
Enhance glob() function to return metadata about matched files.

### DISCUSSION THREAD

**Discussion by rustyconover** [duckdb/duckdb#11823] [Ideas]
I propose that the `glob()` function take an optional named parameter called `metadata` that would add additional columns to `glob()`'s output.  This data would depend on the underlying filesystem, but most typically it would include the last modified date and size of each matched file.

An example call would be:

```sql
select * from glob('*', metadata=True);
```

By including this additional information could be possible to filter the various matched filenames.

**Comment by carlopi** [comment]
There is a:
```
SELECT * EXCLUDE content FROM  read_text('*');
```
That perform basically the same.

Note that `EXCLUDE content` is relevant, since otherwise also the file content is fetched and returned, and that changes performances drastically.

Including this as glob variant could make sense, thanks for the input.

**Comment by rustyconover** [comment]
Actually I'm working with just glob() returning filenames it doesn't have any interaction with the content of the files.

**Reply by carlopi** [reply]
Sure, but if you are looking for extra metadata, read_text() does provide that similarly to the functionality you proposed.
@Maxxen, that might be interested in the glob / read_text mashup.

**Reply by rustyconover** [reply]
Not everything I'm reading is text, sometimes is JSON sometimes its Parquet

**Reply by Mause** [reply]
Try you can try `read_blob` instead? Though I doubt it matters if you're not doing anything with the content

**Reply by rustyconover** [reply]
Mostly it's important to get the file metadata, often we are checking to see if all files are present for a particular pattern.  The file system is the data source not the content itself.

**Reply by Maxxen** [reply]
Both `read_file` and `read_blob` functions basically identically to `glob()` if you don't include the `content` column. Because they use "projection pushdown", If you don't query some of the columns they produce those columns won't get computed, so e.g. `SELECT filename, last_modified, size FROM read_blob('*');` will only fetch the filename, last modified and file size from the underlying file system, it won't actually open and read the files.

we could of course enhance glob to also provide last_modified and size, but as it stands you can already do exactly what glob does + additional metadata with both read_text and read_blob.

**Comment by samansmink** [comment]
@rustyconover I like this idea. I'm atm working on https://github.com/duckdb/duckdb/pull/11806 which implements a refactor of the MultiFileReader introducing the `MultiFileList` concept. We could use the concept of a MultiFileList to also include any potential file metadata such as the filesize, which would be returned by a glob operation. 

This could then be used to avoid the super-lame HEAD request that we do on each file to check its size. Also expose things like cache tags and last modified dates which could be useful during query execution if we even implement FS level caching.
